src/features/harmful-apis.js

```diff
@@ -0,0 +1,443 @@
+ * @param {string[]} permissions permission names to auto-deny
+ */
+function filterPermissionQuery (permissions) {
+    if (!permissions || permissions.length === 0) {
+        return
+    }
+    wrapMethod(globalThis.Permissions.prototype, 'query', async function (nativeImpl, queryObject) {
+        // call the original function first in case it throws an error
+        const origResult = await nativeImpl.call(this, queryObject)
+
+        if (permissions.includes(queryObject.name)) {
+            return {
+                name: queryObject.name,
+                state: 'denied',
+                status: 'denied'
+            }
+        }
+        return origResult
+    })
+}
+
+/**
+ * Blocks some privacy harmful APIs.
+ */
+export default class HarmfulApis extends ContentFeature {
+    init (args) {
+        console.log('INIT! from harmfulAPIs', args)
+        // @ts-expect-error linting is not yet seet up for worker context
+        /** @type Navigator | WorkerNavigator */
+        this.navigatorPrototype = globalThis.Navigator?.prototype || globalThis.WorkerNavigator?.prototype
+
+        this.removeDeviceOrientationEvents(this.getFeatureSetting('deviceOrientation'))
+        this.blockGenericSensorApi(this.getFeatureSetting('GenericSensor'))
+        this.filterUAClientHints(this.getFeatureSetting('UaClientHints'))
+        this.removeNetworkInformationApi(this.getFeatureSetting('NetworkInformation'))
+        this.blockGetInstalledRelatedApps(this.getFeatureSetting('getInstalledRelatedApps'))
+        this.removeFileSystemAccessApi(this.getFeatureSetting('FileSystemAccess'))
+        this.blockWindowPlacementApi(this.getFeatureSetting('WindowPlacement'))
+        this.blockWebBluetoothApi(this.getFeatureSetting('WebBluetooth'))
+        this.blockWebUsbApi(this.getFeatureSetting('WebUsb'))
+        this.blockWebSerialApi(this.getFeatureSetting('WebSerial'))
+        this.blockWebHidApi(this.getFeatureSetting('WebHid'))
+        this.blockWebMidiApi(this.getFeatureSetting('WebMidi'))
+        this.removeIdleDetectionApi(this.getFeatureSetting('IdleDetection'))
+        this.removeWebNfcApi(this.getFeatureSetting('WebNfc'))
+        this.filterStorageManagerApi(this.getFeatureSetting('StorageManager'))
+    }
+
+    /**
+     * @param {DeviceOrientationConfig} settings
+     */
+    removeDeviceOrientationEvents (settings) {
+        if (!settings.protected) {
+            return
+        }
+        const eventsToBlock = settings.filterEvents || []
+        if (eventsToBlock.length > 0) {
+            for (const eventName of eventsToBlock) {
+                const dom0HandlerName = `on${eventName}`
+                if (dom0HandlerName in globalThis) {
+                    delete globalThis[dom0HandlerName]
+                }
+            }
+            wrapMethod(EventTarget.prototype, 'addEventListener', function (nativeImpl, type, ...restArgs) {
```

Comment:
Double wrapping things has been an issue before for us. I would prefer we didn't but also I don't really want to refactor mid PR.

Are you at all concerned by it? (performance and page detectability wise).

This API is called a lot so we should add a lot more scrutiny to the review here.

Reply:
Could you elaborate what kind of issues you had with double wrapping?
Performance is not my biggest concern because we are adding overhead on `addEventListener()` itself, not when the handler is called.
Page detectability is just the same as any other wrapping we do: there are certainly ways to detect this, but you'll need to jump through some hoops.

Reply:
document.cookie was much more complex and had async bits.... so I think we can ignore. It just gives me the ick but as I can't really articulate why it's fine :)

Reply:
I feel you 😅 It is a very common API indeed, so we gotta be careful, but I also couldn't think of any critical flaw of this approach